### PR TITLE
Simplify the RootNavigator structure

### PR DESCRIPTION
**Pull request by WojtekBoman**
### Details

This PR refactors an approach of storing CentralPane screens in the navigation stack. In the old approach all screens displayed in CentralPane are wrapped by `CentralPaneNavigator`, in this PR this wrapper has been removed, screens are stored directly in the navigation stack.

### Fixed Issues

$ https://github.com/Expensify/App/issues/42507
PROPOSAL:

### Tests

- [ ] Verify that no errors appear in the JS console
- [ ] Verify if central pane screens are displayed correctly and animations and deeplinks work fine
Pages to test:
-  SCREENS.SETTINGS.WORKSPACES,
-  SCREENS.SETTINGS.PREFERENCES.ROOT,
-  SCREENS.SETTINGS.SECURITY,
-  SCREENS.SETTINGS.PROFILE.ROOT,
-  SCREENS.SETTINGS.WALLET.ROOT,
-  SCREENS.SETTINGS.ABOUT,
-  SCREENS.SETTINGS.TROUBLESHOOT,
-  SCREENS.SETTINGS.SAVE_THE_WORLD,
-  SCREENS.SETTINGS.SUBSCRIPTION.ROOT,
-  SCREENS.SEARCH.CENTRAL_PANE,
-  SCREENS.REPORT

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/47774969/3b29ca62-d58a-42c0-9352-ea2297671c1a

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/47774969/f243caaf-b880-4c45-93a1-09846e6e5cbc

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/47774969/43b4ca06-3319-43a0-8593-ba17ded26155

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/47774969/66aa39a3-3dad-4a8f-b2ed-7b76200c5589

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/47774969/506b6615-f0da-4f0b-aecb-52c4b454890e

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/47774969/557ca56f-f749-41ec-99f3-65318f7f62b3

</details>

### CODE REVIEW

File: src/libs/Navigation/linkTo.ts
Comment (adamgrzybowski) [inline src/libs/Navigation/linkTo.ts:132]: 🧹

File: src/components/withPrepareCentralPaneScreen/index.native.tsx
Comment (adamgrzybowski) [inline src/components/withPrepareCentralPaneScreen/index.native.tsx:5]: We probably need some comments here and in the other file about what preparing means for native and web.

File: src/libs/Navigation/AppNavigator/AuthScreens.tsx
Comment (adamgrzybowski) [inline src/libs/Navigation/AppNavigator/AuthScreens.tsx:289]: I am wondering if this name is good for this structure

**Comment by adamgrzybowski** [comment]
Looks great to me 💪  Good job!

**Comment by mountiny** [comment]
running a build

**Comment by adamgrzybowski** [comment]
@mountiny should work now!

**Comment by adamgrzybowski** [comment]
Conflicts resolved and it works again

**Review by hoangzinh — COMMENTED** [review]
Overall, LGTM. Just a few small concerns. I will try to test this PR today

File: tests/ui/UnreadIndicatorsTest.tsx
Comment (adamgrzybowski) [reply tests/ui/UnreadIndicatorsTest.tsx:623]: I am not entirely sure but the situation before the PR was wrong. There should be two occurrences in the DOM from the beginning. Not sure why it wasn't working. It's probably something related to nesting + mocks.

File: src/components/ScreenWrapper.tsx
Comment (adamgrzybowski) [reply src/components/ScreenWrapper.tsx:97]: Yeah, the navigation object should be typed with all the screens from AuthNavigator | RootStack

**Comment by mountiny** [comment]
@WojtekBoman is this ready for another testing from @hoangzinh ?

**Comment by adamgrzybowski** [comment]
@mountiny @hoangzinh I fixed the problem with blinking. I found a root cause for the `Unable to access submit expense by direct link`. I will continue with the implementation tomorrow. The second reported issue `App shows not found page forever after sign-in with a deep link` may be related but I have to confirm.

**Comment by adamgrzybowski** [comment]
@hoangzinh @mountiny Hey guys. I spent today quite some time trying to debug the second reported issue `Bug 2: App shows not found page forever after sign-in with a deep link`

I haven't figured out anything. But I realised that it's exists on the main as well. So I think it shouldn't block this PR. 
image

Generally speaking, things that happen after logging in seems unreliable. Sometimes it takes long and sometimes we land on the not found page. I managed to get not found screen on the `new.expensify` only once though. 


The rest of the bugs for this PR should be fixed now.

**Comment by mountiny** [comment]
Thanks @hoangzinh @adamgrzybowski

**Review by hoangzinh — APPROVED** [review]
LGTM

**Review by hoangzinh — APPROVED** [review]

**Comment by WojtekBoman** [comment]
@mountiny I solved conflicts in this PR, we can proceed with it

File: src/components/withPrepareCentralPaneScreen/index.native.tsx
Comment (mountiny) [inline src/components/withPrepareCentralPaneScreen/index.native.tsx:7]:
```suggestion
 * This HOC is dependent on the platform. On native platforms, screens that aren't already displayed in the navigation stack should be frozen to prevent unnecessary rendering.
```

File: src/components/withPrepareCentralPaneScreen/index.native.tsx
Comment (mountiny) [inline src/components/withPrepareCentralPaneScreen/index.native.tsx:8]:
```suggestion
 * It's handled this way only on mobile platforms because on the web, more than one screen is displayed in a wide layout, so these screens shouldn't be frozen.
```

File: src/components/withPrepareCentralPaneScreen/index.tsx
Comment (mountiny) [inline src/components/withPrepareCentralPaneScreen/index.tsx:4]:
```suggestion
 * This HOC is dependent on the platform. On native platforms, screens that aren't already displayed in the navigation stack should be frozen to prevent unnecessary rendering.
```

File: src/components/withPrepareCentralPaneScreen/index.tsx
Comment (mountiny) [inline src/components/withPrepareCentralPaneScreen/index.tsx:5]:
```suggestion
 * It's handled this way only on mobile platforms because on the web, more than one screen is displayed in a wide layout, so these screens shouldn't be frozen.
```

File: src/libs/Navigation/AppNavigator/createCustomBottomTabNavigator/BottomTabBar/index.website.tsx
Comment (mountiny) [inline src/libs/Navigation/AppNavigator/createCustomBottomTabNavigator/BottomTabBar/index.website.tsx:52]: Hmm should not this be also negated like [here](https://github.com/Expensify/App/pull/42582/files#diff-5fb2ba266751a0dc98ab04110f3bc0b249f07db75d3156974504ef3b562ed08cR51)?
```suggestion
        if (!!(currentRoute && currentRoute.name !== NAVIGATORS.BOTTOM_TAB_NAVIGATOR && !isCentralPaneName(currentRoute.name)) || Session.isAnonymousUser()) {
```

File: src/pages/settings/InitialSettingsPage.tsx
Comment (mountiny) [reply src/pages/settings/InitialSettingsPage.tsx:328]: +1 since we are checking for `!!activeRoute &&` this is not necessary right?

**Review by mountiny — COMMENTED** [review]
@WojtekBoman left couple comments, can you please address?

**Comment by adamgrzybowski** [comment]
Hi @dominictb! [The test](https://github.com/Expensify/App/pull/42742) you added recently is failing after our changes. I tested this case manually and it all looks fine. I am trying to figure out what is wrong but I can't figure it out. Could you please take a look? 🙏

**Comment by adamgrzybowski** [comment]
It would be great to merge these changes to avoid more conflicts etc. But maybe we could tackle this test later. @mountiny What do you think?

**Comment by mountiny** [comment]
@WojtekBoman @adamgrzybowski sorry there is more conflicts

@dominictb it would be great to get this fixd before merging.

**Comment by WojtekBoman** [comment]
@mountiny Conflicts solved and all checks passed :)

**Review by mountiny — DISMISSED** [review]
These pesky conflicts

**Review by mountiny — APPROVED** [review]
Going to merge this to hopefully prevent any other conflicts but lets be on a look out for regressions

**Comment by roryabraham** [comment]
We discovered a potential problem with this PR when it made some automated UI tests that we're adding in another PR start failing.

### Actions taken

1. Open report A
2. Click link to report B

### Expected result
OpenReport is called once for report A and once for reportB

### Actual result
OpenReport is called twice for reportA and once for report B.

*note:* This likely happens because `ReportScreen` remounts for report A.

**Comment by yuwenmemon** [comment]
This caused the regression here: https://github.com/Expensify/App/issues/44422 ([Discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1719427136231479))

**Pull request merged**
